tower-http/src/classify.rs

```diff
@@ -0,0 +1,316 @@
+
+    /// Classify an error.
+    ///
+    /// Errors are always errors (doh) but sometimes it might be useful to have multiple classes of
+    /// errors. A retry policy might allows retrying some errors and not others.
+    fn classify_error(self, error: &E) -> Self::FailureClass;
+}
+
+/// Result of doing a classification.
+#[derive(Debug)]
+pub enum ClassifiedResponse<FailureClass, ClassifyEos> {
+    /// The response was able to be classified immediately.
+    Ready(Result<(), FailureClass>),
+    /// We have to wait until the end of a streaming response to classify it.
+    RequiresEos(ClassifyEos),
+}
+
+/// A [`ClassifyEos`] type that can be used in [`ClassifyResponse`] implementations that never have
+/// to classify streaming responses.
+///
+/// `NeverClassifyEos` exists only as type. You cannot construct `NeverClassifyEos` values.
+pub struct NeverClassifyEos<T> {
+    _output_ty: PhantomData<fn() -> T>,
+    _never: Infallible,
+}
+
+impl<T, E> ClassifyEos<E> for NeverClassifyEos<T> {
+    type FailureClass = T;
+
+    fn classify_eos(self, _trailers: Option<&HeaderMap>) -> Result<(), Self::FailureClass> {
+        // `NeverClassifyEos` contains an `Infallible` so it can never be constructed
+        unreachable!()
+    }
+
+    fn classify_error(self, _error: &E) -> Self::FailureClass {
+        // `NeverClassifyEos` contains an `Infallible` so it can never be constructed
+        unreachable!()
+    }
+}
+
+/// The default classifier used for normal HTTP responses.
+///
+/// Responses with a `5xx` status code are failures, all others as successes.
+#[derive(Clone, Debug, Default)]
+pub struct ServerErrorsAsFailures {
+    _priv: (),
+}
+
+impl ServerErrorsAsFailures {
+    /// Create a new [`ServerErrorsAsFailures`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a [`MakeClassifier`] that produces `ServerErrorsAsFailures`.
+    ///
+    /// This is a convenience function that simply calls `SharedClassifier::new`.
+    pub fn make_classifier<E>() -> SharedClassifier<Self> {
+        SharedClassifier::new::<E>(Self::new())
+    }
+}
+
+impl<E> ClassifyResponse<E> for ServerErrorsAsFailures {
+    type FailureClass = StatusCode;
+    type ClassifyEos = NeverClassifyEos<StatusCode>;
+
+    fn classify_response<B>(
+        self,
+        res: &Response<B>,
+    ) -> ClassifiedResponse<Self::FailureClass, Self::ClassifyEos> {
+        if res.status().is_server_error() {
+            ClassifiedResponse::Ready(Err(res.status()))
+        } else {
+            ClassifiedResponse::Ready(Ok(()))
+        }
+    }
+
+    fn classify_error(self, _error: &E) -> Self::FailureClass {
+        StatusCode::INTERNAL_SERVER_ERROR
+    }
+}
+
+/// Response classifier for gRPC responses.
+///
+/// gRPC doesn't use normal HTTP statuses for indicating success or failure but instead a special
+/// header that might appear in a trailer.
+///
+/// Responses where `grpc-status` is 0 are successes, all others are failures.
```

Comment:
This implies that this classifies responses where there is no `grpc-status` trailer as failures, but looking at the implementation, it looks like they are considered successes: https://github.com/tower-rs/tower-http/blob/a220f3768e0f0dcc8798068992eb54f8f7c57450/tower-http/src/classify.rs#L237

I do think this is probably the correct behavior, so let's not change that; we should just make this a little clearer in the docs.

Reply:
Good point. I made it a bit more clear in 4e9f328046de1